bitcoin/src/bip32.rs

```diff
@@ -730,33 +1085,75 @@ impl Xpub {
     /// the internal public key representation.
     pub fn to_x_only_pub(self) -> XOnlyPublicKey { XOnlyPublicKey::from(self.public_key) }
 
+    /// Derive an extended public key from a path that contains NormalChildNumbers.
+    pub fn derive_pub<
+        'a,
+        C: secp256k1::Verification,
+        T: IntoIterator<Item = &'a NormalChildNumber>,
+    >(
+        &self,
+        secp: &Secp256k1<C>,
+        path: T,
+    ) -> Xpub {
+        let mut pk: Xpub = *self;
+        for normal_cnum in path.into_iter() {
+            pk = pk
+                .internal_derive_pub(secp, normal_cnum.to_number())
+                .expect("Unreachable: NormalChildNumber is always valid");
```

Comment:
This doesnt work;
`missing match arm: `CannotDeriveFromHardenedKey`, `Secp256k1(_)`, `InvalidChildNumber(_)` and 8 more not covered`
this `expect()` does the job.

If you want something shorter `unwrap()` does the job too.

Reply:
That proves you did something wrong since `<NormalChildNumber as TryInto<NormalChildNumber>::Error = core::convert::Infallible`. It's not about being short, it's about having a proof it's correct. If you can't figure it out until tomorrow let me know and I'll take a look.

Reply:
Problably, i can handle more attention 
`internal_derive_pub`,
`try_derive_pub` and
`derive_pub`

Sure that internal_derive_pub can be infallible too.


but The `.unwrap_or_else(|never| match never {})`
doesnt work because of the return types that `never` can be, such as all of the `bip32::Error`s... meaning that we could rely on just making `internal_derive_pub` a infallible function. to solve this.

Reply:
Note that bip32 derivation can fail according to the spec.
(if the resulting point is the identity iirc)

Reply:
`internal_derive_pub` must take `C: TryInto<ChildNumber>` and return `Result<_, C::Error>`. Then `C::Error` is `Infallible` for `ChildNumber` therefore the match must work. If it doesn't you did something wrong. If you can't figure it out, post the code for me to look at it.

Reply:
Why not make it infallible ?

Reply:
Because converting everything to `NormalDerivationPath` has negative impact on performance.

I also have different design idea in mind but that is complex in different ways and might require `unsafe` to get full performance.

Reply:
Ok, will make a review tagging the code so you can take a look into it.